system-tests/e2e-testsupport/src/main/java/com/hotels/styx/servers/MockOriginServer.java

```diff
@@ -0,0 +1,218 @@
+import com.github.tomakehurst.wiremock.http.Request;
+import com.github.tomakehurst.wiremock.http.RequestHandler;
+import com.github.tomakehurst.wiremock.http.StubRequestHandler;
+import com.github.tomakehurst.wiremock.http.StubResponseRenderer;
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.ServiceManager;
+import com.hotels.styx.api.HttpHandler2;
+import com.hotels.styx.server.HttpConnectorConfig;
+import com.hotels.styx.server.HttpServer;
+import com.hotels.styx.server.HttpServers;
+import com.hotels.styx.server.HttpsConnectorConfig;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static com.github.tomakehurst.wiremock.client.WireMock.configureFor;
+import static com.github.tomakehurst.wiremock.client.WireMock.stubFor;
+import static com.google.common.base.Optional.absent;
+import static com.hotels.styx.servers.WiremockResponseConverter.toStyxResponse;
+import static java.lang.String.format;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Collections.emptyMap;
+import static rx.Observable.just;
+
+public final class MockOriginServer {
+    private static final String FILES_ROOT = "__files";
+    private static final Logger LOGGER = LoggerFactory.getLogger(MockOriginServer.class);
+    private static final int MAX_CONTENT_LENGTH = 256 * 1024;
+
+    private final String appId;
+    private final String originId;
+    private final int adminPort;
+
+    private int serverPort;
+    private final HttpServer adminServer;
+    private final HttpServer mockServer;
+
+    static {
+        System.setProperty("org.mortbay.log.class", "com.github.tomakehurst.wiremock.jetty.LoggerAdapter");
+    }
+
+    private ServiceManager services;
+
+    private MockOriginServer(String appId, String originId, int adminPort, int serverPort, HttpServer adminServer, HttpServer mockServer) {
+        this.appId = appId;
+        this.originId = originId;
+        this.adminPort = adminPort;
+        this.serverPort = serverPort;
+        this.adminServer = adminServer;
+        this.mockServer = mockServer;
+    }
+
+    private static WireMockApp wireMockApp() {
+        return new WireMockApp(
+                new NotImplementedRequestDelayControl(),
+                false,
+                stubMappings -> {
+
+                },
+                mappings -> {
+
+                },
+                false,
+                absent(),
+                emptyMap(),
+                null,
+                null
+        );
+    }
+
+    private static HttpServer createAdminServer(String originId, int adminPort, WireMockApp wireMockApp) {
+        return HttpServers.createHttpServer(
+                "mock-admin-" + originId,
+                new HttpConnectorConfig(adminPort),
+                adminHandler(wireMockApp));
+    }
+
+    public static MockOriginServer create(String appId, String originId, int adminPort, HttpConnectorConfig httpConfig) {
```

Comment:
There is a lot of code duplication between the two `create` methods - is it possible to neatly refactor them to reduce this?

Reply:
The bulkiest part of the code is in the middle: a call to static `createHttpServer`, or `createHttpsServer`. To have a meaningful outcome we should look into refactoring those two static methods instead.

Personally I am happy with this code as it is, as I need this PR in for additional TLS related features. I would refactor in future once we'll have a more pressing reason to do so. But please feel free to contribute if you have some good ideas.